### PR TITLE
Port Microsoft.Extensions.[A..D]

### PR DESCRIPTION
**Pull request by carlossanlop**
Caching, Compliance, Configuration, DependencyInjection
Please review ignoring whitespace

### CODE REVIEW

File: xml/Microsoft.Extensions.DependencyInjection/SocketsHttpHandlerBuilderExtensions.xml [hunk 1]
```diff
@@ -1,98 +1,103 @@
...
+      </Parameters>
+      <Docs>
+        <param name="builder">The <see cref="T:Microsoft.Extensions.DependencyInjection.ISocketsHttpHandlerBuilder" />.</param>
+        <param name="configuration">Configuration containing properties of <see cref="T:System.Net.Http.SocketsHttpHandler" />.</param>
+        <summary>Uses <see cref="T:Microsoft.Extensions.Configuration.IConfiguration" /> to configure the primary <see cref="T:System.Net.Http.SocketsHttpHandler" /> for a named <see cref="T:System.Net.Http.HttpClient" />.</summary>
+        <returns>An <see cref="T:Microsoft.Extensions.DependencyInjection.ISocketsHttpHandlerBuilder" /> that can be used to configure the handler.</returns>
+        <remarks>
+          <para> Only simple (of type `bool`, `int`, <see cref="T:System.Enum" /> or <see cref="T:System.TimeSpan" />) properties of <see cref="T:System.Net.Http.SocketsHttpHandler" /> will be parsed.
+            </para>
+          <para> All unmatched properties in <see cref="T:Microsoft.Extensions.Configuration.IConfiguration" /> will be ignored.
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/SocketsHttpHandlerBuilderExtensions.xml:64]:
```suggestion
          <para> All unmatched properties in <see cref="T:Microsoft.Extensions.Configuration.IConfiguration" /> are ignored.
```

File: xml/Microsoft.Extensions.DependencyInjection/SocketsHttpHandlerBuilderExtensions.xml [hunk 2]
```diff
@@ -1,98 +1,103 @@
...
+        <Parameter Name="builder" Type="Microsoft.Extensions.DependencyInjection.ISocketsHttpHandlerBuilder" RefType="this" />
+        <Parameter Name="configuration" Type="Microsoft.Extensions.Configuration.IConfiguration" />
+      </Parameters>
+      <Docs>
+        <param name="builder">The <see cref="T:Microsoft.Extensions.DependencyInjection.ISocketsHttpHandlerBuilder" />.</param>
+        <param name="configuration">Configuration containing properties of <see cref="T:System.Net.Http.SocketsHttpHandler" />.</param>
+        <summary>Uses <see cref="T:Microsoft.Extensions.Configuration.IConfiguration" /> to configure the primary <see cref="T:System.Net.Http.SocketsHttpHandler" /> for a named <see cref="T:System.Net.Http.HttpClient" />.</summary>
+        <returns>An <see cref="T:Microsoft.Extensions.DependencyInjection.ISocketsHttpHandlerBuilder" /> that can be used to configure the handler.</returns>
+        <remarks>
+          <para> Only simple (of type `bool`, `int`, <see cref="T:System.Enum" /> or <see cref="T:System.TimeSpan" />) properties of <see cref="T:System.Net.Http.SocketsHttpHandler" /> will be parsed.
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/SocketsHttpHandlerBuilderExtensions.xml:62]:
```suggestion
          <para>Only properties of <see cref="T:System.Net.Http.SocketsHttpHandler" /> of type <c>bool</c>, <c>int</c>, <see cref="T:System.Enum" />, or <see cref="T:System.TimeSpan" /> are parsed.
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml [hunk 3]
```diff
@@ -1,243 +1,245 @@
...
+      <AttributeName Language="C#">[System.Runtime.CompilerServices.Nullable(0)]</AttributeName>
+      <AttributeName Language="F#">[&lt;System.Runtime.CompilerServices.Nullable(0)&gt;]</AttributeName>
+    </Attribute>
+    <Attribute>
+      <AttributeName Language="C#">[System.Runtime.CompilerServices.NullableContext(1)]</AttributeName>
+      <AttributeName Language="F#">[&lt;System.Runtime.CompilerServices.NullableContext(1)&gt;]</AttributeName>
+    </Attribute>
+  </Attributes>
+  <Docs>
+    <summary>Extension methods for getting services from an <see cref="T:System.IServiceProvider" />.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml:27]:
```suggestion
    <summary>Provides extension methods for getting services from an <see cref="T:System.IServiceProvider" />.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml [hunk 4]
```diff
@@ -1,243 +1,245 @@
...
+            </Attribute>
+          </Attributes>
+        </Parameter>
+        <Parameter Name="serviceKey" Type="System.Object" />
+      </Parameters>
+      <Docs>
+        <typeparam name="T">The type of service object to get.</typeparam>
+        <param name="provider">The <see cref="T:System.IServiceProvider" /> to retrieve the service object from.</param>
+        <param name="serviceKey">An object that specifies the key of service object to get.</param>
+        <summary>Get service of type <typeparamref name="T" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml:70]:
```suggestion
        <summary>Gets a service of type <typeparamref name="T" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml [hunk 5]
```diff
@@ -1,243 +1,245 @@
...
+              <AttributeName Language="F#">[&lt;System.Runtime.CompilerServices.Nullable(2)&gt;]</AttributeName>
+            </Attribute>
+          </Attributes>
+        </Parameter>
+      </Parameters>
+      <Docs>
+        <param name="provider">The <see cref="T:System.IServiceProvider" /> to retrieve the services from.</param>
+        <param name="serviceType">An object that specifies the type of service object to get.</param>
+        <param name="serviceKey">An object that specifies the key of service object to get.</param>
+        <summary>Get an enumeration of services of type <paramref name="serviceType" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml:118]:
```suggestion
        <summary>Gets an enumeration of services of type <paramref name="serviceType" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml [hunk 6]
```diff
@@ -1,243 +1,245 @@
...
+              <AttributeName Language="F#">[&lt;System.Runtime.CompilerServices.Nullable(2)&gt;]</AttributeName>
+            </Attribute>
+          </Attributes>
+        </Parameter>
+      </Parameters>
+      <Docs>
+        <typeparam name="T">The type of service object to get.</typeparam>
+        <param name="provider">The <see cref="T:System.IServiceProvider" /> to retrieve the services from.</param>
+        <param name="serviceKey">An object that specifies the key of service object to get.</param>
+        <summary>Get an enumeration of services of type <typeparamref name="T" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml:163]:
```suggestion
        <summary>Gets an enumeration of services of type <typeparamref name="T" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml [hunk 7]
```diff
@@ -1,243 +1,245 @@
...
+              <AttributeName Language="F#">[&lt;System.Runtime.CompilerServices.Nullable(2)&gt;]</AttributeName>
+            </Attribute>
+          </Attributes>
+        </Parameter>
+      </Parameters>
+      <Docs>
+        <param name="provider">The <see cref="T:System.IServiceProvider" /> to retrieve the service object from.</param>
+        <param name="serviceType">An object that specifies the type of service object to get.</param>
+        <param name="serviceKey">An object that specifies the key of service object to get.</param>
+        <summary>Get service of type <paramref name="serviceType" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml:199]:
```suggestion
        <summary>Gets a service of type <paramref name="serviceType" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml [hunk 8]
```diff
@@ -1,243 +1,245 @@
...
+              <AttributeName Language="F#">[&lt;System.Runtime.CompilerServices.Nullable(2)&gt;]</AttributeName>
+            </Attribute>
+          </Attributes>
+        </Parameter>
+      </Parameters>
+      <Docs>
+        <typeparam name="T">The type of service object to get.</typeparam>
+        <param name="provider">The <see cref="T:System.IServiceProvider" /> to retrieve the service object from.</param>
+        <param name="serviceKey">An object that specifies the key of service object to get.</param>
+        <summary>Get service of type <typeparamref name="T" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProviderKeyedServiceExtensions.xml:238]:
```suggestion
        <summary>Gets a service of type <typeparamref name="T" /> from the <see cref="T:System.IServiceProvider" />.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceKeyAttribute.xml [hunk 9]
```diff
@@ -1,45 +1,45 @@
...
+  </Base>
+  <Interfaces />
+  <Attributes>
+    <Attribute>
+      <AttributeName Language="C#">[System.AttributeUsage(System.AttributeTargets.Parameter)]</AttributeName>
+      <AttributeName Language="F#">[&lt;System.AttributeUsage(System.AttributeTargets.Parameter)&gt;]</AttributeName>
+    </Attribute>
+  </Attributes>
+  <Docs>
+    <summary>ServiceKeyAttribute can be specified on a parameter to inject the key that was used for registration/resolution.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceKeyAttribute.xml:23]:
```suggestion
    <summary>Specifies the parameter to inject the key that was used for registration or resolution.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProvider.xml [hunk 10]
```diff
@@ -1,253 +1,254 @@
...
+          </Attributes>
+        </Parameter>
+      </Parameters>
+      <Docs>
+        <param name="serviceType">The type of the service to get.</param>
+        <param name="serviceKey">The key of the service to get.</param>
+        <summary>Gets the service object of the specified type. Will throw if the service not found.</summary>
+        <returns>The keyed service.</returns>
+        <remarks>To be added.</remarks>
+        <exception cref="T:System.InvalidOperationException" />
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProvider.xml:213]:
```suggestion
        <exception cref="T:System.InvalidOperationException">The service wasn't found.</exception>
```

File: xml/Microsoft.Extensions.DependencyInjection/ServiceProvider.xml [hunk 11]
```diff
@@ -1,253 +1,254 @@
...
+              <AttributeName Language="C#">[System.Runtime.CompilerServices.Nullable(2)]</AttributeName>
+              <AttributeName Language="F#">[&lt;System.Runtime.CompilerServices.Nullable(2)&gt;]</AttributeName>
+            </Attribute>
+          </Attributes>
+        </Parameter>
+      </Parameters>
+      <Docs>
+        <param name="serviceType">The type of the service to get.</param>
+        <param name="serviceKey">The key of the service to get.</param>
+        <summary>Gets the service object of the specified type. Will throw if the service not found.</summary>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/ServiceProvider.xml:210]:
```suggestion
        <summary>Gets the service object of the specified type.</summary>
```

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml [hunk 12]
```diff
@@ -1,1033 +1,1060 @@
...
+      </ReturnValue>
+      <Parameters>
+        <Parameter Name="builder" Type="Microsoft.Extensions.DependencyInjection.IHttpClientBuilder" Index="0" FrameworkAlternate="dotnet-plat-ext-8.0" RefType="this" />
+        <Parameter Name="httpClientLoggerFactory" Type="System.Func&lt;System.IServiceProvider,Microsoft.Extensions.Http.Logging.IHttpClientLogger&gt;" Index="1" FrameworkAlternate="dotnet-plat-ext-8.0" />
+        <Parameter Name="wrapHandlersPipeline" Type="System.Boolean" Index="2" FrameworkAlternate="dotnet-plat-ext-8.0" />
+      </Parameters>
+      <Docs>
+        <param name="builder">The <see cref="T:Microsoft.Extensions.DependencyInjection.IHttpClientBuilder" />.</param>
+        <param name="httpClientLoggerFactory">A delegate that is used to create a custom logger. The logger should implement <see cref="T:Microsoft.Extensions.Http.Logging.IHttpClientLogger" /> or <see cref="T:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger" />.</param>
+        <param name="wrapHandlersPipeline">Whether the logging handler with the custom logger would be added to the top or to the bottom of the additional handlers chains.</param>
```
[the remark this answers is not in the document]
Comment (gewarren) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:216]: Really not sure if this is a correct interpretation.

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml [hunk 13]
```diff
@@ -1,1033 +1,1060 @@
...
+        </TypeParameter>
+      </TypeParameters>
+      <Parameters>
+        <Parameter Name="builder" Type="Microsoft.Extensions.DependencyInjection.IHttpClientBuilder" Index="0" FrameworkAlternate="dotnet-plat-ext-8.0" RefType="this" />
+        <Parameter Name="wrapHandlersPipeline" Type="System.Boolean" Index="1" FrameworkAlternate="dotnet-plat-ext-8.0" />
+      </Parameters>
+      <Docs>
+        <typeparam name="TLogger">The service type of the custom logger as it was registered in DI. The logger should implement <see cref="T:Microsoft.Extensions.Http.Logging.IHttpClientLogger" /> or <see cref="T:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger" />.</typeparam>
+        <param name="builder">The <see cref="T:Microsoft.Extensions.DependencyInjection.IHttpClientBuilder" />.</param>
+        <param name="wrapHandlersPipeline">Whether the logging handler with the custom logger would be added to the top or to the bottom of the additional handlers chains.</param>
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:265]:
```suggestion
        <param name="wrapHandlersPipeline"><see langword="true" /> to add the logging handler with the custom logger to the top of the additional handlers chain; <see langword="false" /> to add it to the bottom of the chain.</param>
```

Comment (gewarren) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:265]: Again, not sure if correct.

**Review by gewarren — APPROVED** [review]

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml [hunk 14]
```diff
@@ -1,1033 +1,1060 @@
...
+        <returns>An <see cref="T:Microsoft.Extensions.DependencyInjection.IHttpClientBuilder" /> that can be used to configure the client.</returns>
+        <remarks>
+          <para> If the <see paramref="wrapHandlersPipeline" /> is `true`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed before all other additional handlers in the chain. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed after all other additional handlers, essentially wrapping the whole pipeline.
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:220]:
```suggestion
          <para> If <paramref name="wrapHandlersPipeline" /> is `true`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed before all other additional handlers in the chain. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed after all other additional handlers, essentially wrapping the whole pipeline.
```

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml [hunk 15]
```diff
@@ -1,1033 +1,1060 @@
...
+            </para>
+          <para> If the <see paramref="wrapHandlersPipeline" /> is `false`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed after all other additional handlers in the chain, right before the primary handler. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed before all other additional handlers, right after the primary handler.
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:222]:
```suggestion
          <para> If <paramref name="wrapHandlersPipeline" /> is `false`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed after all other additional handlers in the chain, right before the primary handler. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed before all other additional handlers, right after the primary handler.
```

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml [hunk 16]
```diff
@@ -1,1033 +1,1060 @@
...
+        <returns>An <see cref="T:Microsoft.Extensions.DependencyInjection.IHttpClientBuilder" /> that can be used to configure the client.</returns>
+        <remarks>
+          <para> If the <see paramref="wrapHandlersPipeline" /> is `true`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed before all other additional handlers in the chain. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed after all other additional handlers, essentially wrapping the whole pipeline.
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:269]:
```suggestion
          <para> If <paramref name="wrapHandlersPipeline" /> is `true`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed before all other additional handlers in the chain. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed after all other additional handlers, essentially wrapping the whole pipeline.
```

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml [hunk 17]
```diff
@@ -1,1033 +1,1060 @@
...
+            </para>
+          <para> If the <see paramref="wrapHandlersPipeline" /> is `false`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed after all other additional handlers in the chain, right before the primary handler. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed before all other additional handlers, right after the primary handler.
```
Comment (gewarren) [inline xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:271]:
```suggestion
          <para> If <paramref name="wrapHandlersPipeline" /> is `false`, <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStart(System.Net.Http.HttpRequestMessage)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStartAsync(System.Net.Http.HttpRequestMessage,System.Threading.CancellationToken)" /> would be executed after all other additional handlers in the chain, right before the primary handler. <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientLogger.LogRequestStop(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan)" /> and <see cref="M:Microsoft.Extensions.Http.Logging.IHttpClientAsyncLogger.LogRequestStopAsync(System.Object,System.Net.Http.HttpRequestMessage,System.Net.Http.HttpResponseMessage,System.TimeSpan,System.Threading.CancellationToken)" /> would be executed before all other additional handlers, right after the primary handler.
```

**Review by gewarren — COMMENTED** [review]
A few more (to fix the build suggestions).

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (carlossanlop) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:216]: @dotnet/area-extensions-dependencyinjection  can you please confirm?

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml (hunk 13 above)
Comment (carlossanlop) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:265]: @dotnet/area-extensions-dependencyinjection can you please confirm?

Comment (carlossanlop) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:265]: Confirmed:
https://github.com/dotnet/runtime/blob/b41b63973bf49ad7a2d0d2ef37f08cf5df622ea1/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.Logging.cs#L58-L65

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (carlossanlop) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:216]: Confirmed:
https://github.com/dotnet/runtime/blob/b41b63973bf49ad7a2d0d2ef37f08cf5df622ea1/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.Logging.cs#L58-L65

File: xml/Microsoft.Extensions.DependencyInjection/ActivatorUtilities.xml [hunk 18]
```diff
@@ -1,340 +1,340 @@
...
+        </TypeParameter>
+      </TypeParameters>
+      <Parameters>
+        <Parameter Name="argumentTypes" Type="System.Type[]" Index="0" FrameworkAlternate="dotnet-plat-ext-8.0" />
+      </Parameters>
+      <Docs>
+        <typeparam name="T">The type to activate</typeparam>
+        <param name="argumentTypes">The types of objects, in order, that will be passed to the returned function as its second parameter.</param>
+        <summary>Creates a delegate that will instantiate a type with constructor arguments provided directly or from an <see cref="T:System.IServiceProvider" />.</summary>
+        <returns>A factory that will instantiate type <paramref name="T" /> using an <see cref="T:System.IServiceProvider" /> and an argument array containing objects matching the types defined in <paramref name="argumentTypes">.</returns>
```
Comment (carlossanlop) [inline xml/Microsoft.Extensions.DependencyInjection/ActivatorUtilities.xml:127]:
```suggestion
        <typeparam name="T">The type to activate.</typeparam>
        <param name="argumentTypes">The types of objects, in order, that will be passed to the returned function as its second parameter.</param>
        <summary>Creates a delegate that will instantiate a type with constructor arguments provided directly or from an <see cref="T:System.IServiceProvider" />.</summary>
        <returns>A factory that will instantiate type <paramref name="T" /> using an <see cref="T:System.IServiceProvider" /> and an argument array containing objects matching the types defined in <paramref name="argumentTypes"/>.</returns>

```

File: xml/Microsoft.Extensions.DependencyInjection/IKeyedServiceProvider.xml [hunk 19]
```diff
@@ -1,103 +1,106 @@
...
+            </Attribute>
+          </Attributes>
+        </Parameter>
+        <Parameter Name="serviceKey" Type="System.Object" />
+      </Parameters>
+      <Docs>
+        <param name="serviceType">An object that specifies the type of service object to get.</param>
+        <param name="serviceKey">An object that specifies the key of service object to get.</param>
+        <summary>Gets the service object of the specified type.</summary>
+        <returns>A service object of type <paramref name="serviceType">, or <see langword="null" /> if there is no service object of type <paramref name="serviceType">.</returns>
```
Comment (carlossanlop) [inline xml/Microsoft.Extensions.DependencyInjection/IKeyedServiceProvider.xml:59]:
```suggestion
        <returns>A service object of type <paramref name="serviceType"/>, or <see langword="null" /> if there is no service object of type <paramref name="serviceType"/>.</returns>

```

File: xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml (hunk 13 above)
Comment (gewarren) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml:265]: @carlossanlop What do you think about the terms top and bottom? Do you think beginning and end are better? I do.

Comment (carlossanlop) [reply xml/Microsoft.Extensions.DependencyInjection/HttpClientBuilderExtensions.xml]: I think the consumer would know what they refer to. We can always address these docs later.

**Pull request merged**